src/orng.cpp

```diff
@@ -309,18 +312,127 @@ void orng::_reward_oracles(asset qty) {
+            s.last_accrue = current_time_point();
+            s.usd_accrued = 0;
+        });
+    } else {
+        // Update existing entry - only update active flag, preserve accrued amount
+        ostip_t.modify(stip_itr, same_payer, [&](auto& s) {
+            s.active = active;
+        });
+    }
+}
+
+/* Accrue stipend for an oracle up to the given time point */
+uint64_t orng::_accrue_stipend(name oracle, time_point now) {
+    ostip_table_type ostip_t(get_self(), get_self().value);
+    auto stip_itr = ostip_t.find(oracle.value);
+
+    if (stip_itr == ostip_t.end()) {
+        return 0;  // No stipend entry, nothing to accrue
+    }
+
+    // Only accrue if oracle is active
+    if (!stip_itr->active) {
+        return 0;
+    }
+
+    // Calculate time delta in seconds
+    int64_t delta_seconds = now.sec_since_epoch() - stip_itr->last_accrue.sec_since_epoch();
+    uint64_t stipend_per_month = get_config(stipendmonth_index, 0);
+
+    // Calculate accrued stipend: (delta_seconds * stipend_per_month) / (30 * 24 * 3600)
+    // Use 30 days = 2,592,000 seconds as the monthly period
+    const uint64_t SECONDS_PER_MONTH = 30 * 24 * 3600;  // 2,592,000
+    uint64_t accrued = (delta_seconds * stipend_per_month) / SECONDS_PER_MONTH;
+
+    return accrued;
+}
+
 void orng::claim(const eosio::name& oracle) {
     eosio::check(!is_paused(), "paused");
     require_auth(oracle);
+
+    auto now = current_time_point();
+
+    // Get stipend entry (oracle must exist in ostip table)
+    ostip_table_type ostip_t(get_self(), get_self().value);
+    auto stip_itr = ostip_t.find(oracle.value);
+    check(stip_itr != ostip_t.end(), "not an oracle");
+
+    // Throttle claims - enforce minimum claim interval
+    uint64_t min_interval = get_config(minclaimint_index, 86400);  // default 24 hours
+    check((now - stip_itr->last_claim).to_seconds() >= min_interval, "too soon");
+
+    // Accrue stipend to 'now' (only if active)
+    uint64_t accrued = _accrue_stipend(oracle, now);
+
+    // Update stipend state with accrued amount
+    if (accrued > 0) {
+        ostip_t.modify(stip_itr, same_payer, [&](auto& s) {
+            s.usd_accrued += accrued;
+            s.last_accrue = now;
+        });
+    }
+
+    // Fetch WAX/USD price from Delphi oracle
+    uint64_t px = _fetch_wax_usd_price();  // BASE_PRECISION format
+
+    // Convert stipend USD → WAX
+    asset wax_from_stip = _usd_to_wax(stip_itr->usd_accrued, px);
+
+    // Get per-call fees from balances table
     bal_table_type bt(get_self(), get_self().value);
-    auto it = bt.require_find(oracle.value, "no balance");
-    check(it->unpaid.amount > 0, "zero balance");
-    asset pay = it->unpaid;
-    bt.modify(it, same_payer, [&](auto& r) { r.unpaid.amount = 0; });
+    auto bal_itr = bt.find(oracle.value);
+    asset fees = (bal_itr != bt.end()) ? bal_itr->unpaid : asset(0, WAX);
+
+    // Calculate total due
+    asset total_due = fees + wax_from_stip;
+
```

Comment:
Should we check if the user has anything to claim? It may be that this condition is impossible but it won't hurt.
`check(total_due > 0, "You have nothing to claim");`